scrapfly/api_response.py

```diff
@@ -565,3 +568,309 @@ def sink(self, path: Optional[str] = None, name: Optional[str] = None, file: Opt
+        self.request = request
+        self.response = response
+        self.screenshot_config = screenshot_config
+        self.result = self.handle_api_result(api_result)
+
+    @property
+    def screenshot_result(self) -> Optional[Dict]:
+        return self.result.get('result', None)
+
+    @property
+    def binary(self) -> Optional[str]:
+        if self.screenshot_result is None:
+            return ''
+        
+        return self.screenshot_result['binary']
+    
+    @property
+    def format(self) -> Optional[str]:
+        if self.screenshot_result is None:
+            return ''
+        
+        return self.screenshot_result['format']
+
+    @property
+    def status_code(self) -> int:
+        return self.response.status_code
+
+    @property
+    def remaining_quota(self) -> Optional[int]:
+        remaining_scrape = self.response.headers.get('X-Scrapfly-Remaining-Scrape')
+
+        if remaining_scrape:
+            remaining_scrape = int(remaining_scrape)
+
+        return remaining_scrape
+
+    @property
+    def cost(self) -> Optional[int]:
+        cost = self.response.headers.get('X-Scrapfly-Api-Cost')
+
+        if cost:
+            cost = int(cost)
+
+        return cost
+
+    @property
+    def screenshot_success(self) -> bool:
+        screenshot_result = self.screenshot_result
+        if not screenshot_result:
+            return False
+
+        return True
+
+    @property
+    def error(self) -> Optional[Dict]:
+        if self.screenshot_result is None:
+            return None
+
+        if self.screenshot_success is False:
+            return self.result
+
+    @property
+    def error_message(self):
+        if self.error is not None:
+            message = "<-- %s | %s - %s." % (self.result['http_code'], self.result['code'], self.result['message'])
+
+            if self.result['links']:
+                message += " Checkout the related doc: %s" % list(self.result['links'].values())[0]
+
+            return message
+        
+        message = "<-- %s | %s." % (self.response.status_code, self.result['message'])
+
+        if self.result.get('links'):
+            message += " Checkout the related doc: %s" % ", ".join(self.result['links'])
+
+        return message        
+
+    def prevent_extra_usage(self):
+        if self.remaining_quota == 0:
+            raise ExtraUsageForbidden(
+                message='All Pre Paid Quota Used',
+                code='ERR::ACCOUNT::PREVENT_EXTRA_USAGE',
+                http_status_code=429,
+                is_retryable=False
+            )
+
+    def _is_api_error(self, api_result: Dict) -> bool:
+        if api_result is None:
+            return True
+
+        return 'error_id' in api_result
+
+    def handle_api_result(self, api_result: bytes) -> FrozenDict:
+        try:
+            api_result = msgpack.unpackb(api_result, raw=False)
```

Comment:
This depends on the `accept` header of the client, which can be either json or msgpack. You can use the `ResponseBodyHandler`. Additionally, you import `msgpack`. The ResponseBodyHandler will detect if msgpack is available or not and handle the situation if msgpack is not present, so you must remove the msgpack import.